include/mbgl/util/thread_priority_setter.hpp

```diff
@@ -0,0 +1,38 @@
+#pragma once
+
+#include <mbgl/platform/settings.hpp>
+#include <mbgl/util/platform.hpp>
+
+namespace mbgl {
+namespace util {
+
+// Thread priority strategies
+//
+// ThreadPrioritySetterLow: Sets thread priority to low. Platform must set
+// correct thread priority via implementing platform::makeThreadLowPriority()
+//
+// ThreadPrioritySetterCustom: Sets thread priority for a thread type based on
+// a setting platform::EXPERIMENTAL_THREAD_PRIORITY_* for a provided thread type.
+
+struct ThreadPrioritySetter {};
```

Comment:
why not just `std::function<void()>` ?

Reply:
`std::function` erases type, I need type to do a branch between `ThreadPrioritySetterLow` and `ThreadPrioritySetterCustom`

Reply:
why not check at run-time that function does not point to a target, and if not fallback to low priority?

Reply:
> why not check at run-time that function does not point to a target

Sounds like a hack :D but I can do that. Empty struct or struct with one string would produce less bloat and easier to understand.

Reply:
why hack? you just provide a custom priority setter, and if you don't - it's just using the default priority. Empty struct looks hackier IMO :smile:

Reply:
@pozdnyakov using `std::function::target` is a hack for many reasons. First of all, it requires rtti that we've disabled for core builds. Secondly, even if target is known, it will just return pointer to callable object, how that would be helpful?

Reply:
oh, what i meant is just
```
std::function setter;
if (!setter) use default
```
not fetching `std::function::target` itself :smile:

Reply:
@pozdnyakov switched to std::function in https://github.com/mapbox/mapbox-gl-native/pull/16174/commits/c9b30f471fb1d4aeae7029a280751e5a9a9ee5c9